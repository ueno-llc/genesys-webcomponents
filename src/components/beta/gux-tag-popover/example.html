
<gux-tag-popover-beta position="right">
    <gux-option>American English</gux-option>
    <gux-option>Latin American Spanish</gux-option>
    <gux-option>European Spanish</gux-option>
    <gux-option>UK English</gux-option>
    <gux-option>Canadian French</gux-option>
    <gux-option>European French</gux-option>
    <gux-option>Dutch</gux-option>
</gux-tag-popover-beta>
<br />
<gux-tag-popover-beta color="dark-blue" position="left">
    <gux-option>American English</gux-option>
    <gux-option>Latin American Spanish</gux-option>
    <gux-option>European Spanish</gux-option>
    <gux-option>UK English</gux-option>
    <gux-option>Canadian French</gux-option>
    <gux-option>European French</gux-option>
    <gux-option>Dutch</gux-option>
</gux-tag-popover-beta>
<br />
<gux-tag-popover-beta color="dark-purple" position="left">
    <gux-option>American English</gux-option>
    <gux-option>Latin American Spanish</gux-option>
    <gux-option>European Spanish</gux-option>
    <gux-option>UK English</gux-option>
    <gux-option>Canadian French</gux-option>
    <gux-option>European French</gux-option>
    <gux-option>Dutch</gux-option>
</gux-tag-popover-beta>