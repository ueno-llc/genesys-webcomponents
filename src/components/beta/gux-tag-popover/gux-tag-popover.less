@import (reference) '../../../style/color.less';

// Variables part
@grey: #222529;

@menu-background-color: @gux-off-white;
@menu-border-color: #dae1eb;

@border-radius: 2px;

// Style
gux-tag-popover-beta {
  div.gux-tag-popover {
    position: relative;
    margin: 4px 0;

    &.right {
      .gux-tag-popover-container {
        text-align: right;
      }
      .arrow {
        margin-right: 19px;
        right: 0;
      }
    }

    &.left {
      .gux-tag-popover-container {
        text-align: left;
      }
      .arrow {
        margin-left: 19px;
        left: 0;
      }
    }

    .gux-tag-popover-container {
      position: relative;
      height: 32px;
      width: 100%;

      .gux-tag-popover-button {
        button {
          font-size: 16px;
        }

        &.has-tags > button {
          color: @gux-dark-blue;
        }

        &:focus {
          outline: none;
          box-shadow: 0 0 0 3px fade(@gux-blue, 50%);
        }
      }
    }

    .gux-tag-popover-menu {
      display: none;
      box-shadow: 0 0 2px fade(@grey, 24%);
      margin-top: 8px;
      position: relative;
      width: 100%;

      .arrow,
      .arrow::after {
        position: absolute;
        content: ' ';
        height: 0;
        width: 0;
        border-style: solid;
        border-color: transparent;
      }
      .arrow {
        border-width: 6px;
        border-bottom-color: @gux-border;
        margin-top: -12px;
      }
      .arrow::after {
        border-width: 4px;
        border-bottom-color: @gux-off-white;
        margin-top: -2px;
        margin-left: -4px;
      }

      .gux-tag-popover-menu-container {
        padding: 8px;
        background: @gux-off-white;

        .gux-tag-popover-options-container {
          display: flex;

          .gux-tag-popover-title {
            padding: 0 5px;
            white-space: nowrap;
            display: flex;
            flex-direction: column;
            justify-content: center;
          }

          .gux-tag-popover-chip-container {
            margin-left: 7px;
            padding: 2px 5px;
            display: flex;
            border-radius: @border-radius;
            border: 1px solid @gux-charcoal-grey;
            flex: 1 1 auto;
            flex-wrap: wrap;
            &.active {
              border-color: @gux-dark-blue;
            }
          }

          .gux-tag-popover-chip {
            display: flex;
            margin: 4px 0;
          }

          .gux-tag-popover-chip-text,
          .gux-tag-popover-chip-close-icon-wrap  {
            background-color: @gux-charcoal-grey;
            height: 24px;
            border-radius: @border-radius;
            color: white;
            margin-right: 8px;
            font-weight: bold;
            text-align: center;
          }

          .gux-tag-popover-chip-text {
            display: flex;
            white-space: nowrap;
            padding: 0 8px;
            line-height: 24px;
          }

          .gux-tag-popover-chip:hover {
            .gux-tag-popover-chip-text {
              margin-right: 1px;
              border-radius: @border-radius 0 0 @border-radius;
            }
            .gux-tag-popover-chip-close-icon-wrap {
              display: flex;
              border-radius: 0 @border-radius @border-radius 0;
            }
          }

          .gux-tag-popover-chip-close-icon-wrap {
            width: 24px;
            box-sizing: border-box;
            display: none;
            cursor: pointer;
            flex-direction: column;
            justify-content: center;
          }

          .gux-tag-popover-chip-icon-wrap {
            display: flex;
            flex-direction: column;
            justify-content: center;
            padding-right: 8px;
          }

          .gux-tag-popover-chip-close-icon {
            margin: 0 auto;
          }

        // GUX-DROPDOWN component
          gux-dropdown input {
            height: 24px;
            background-color: transparent;
            padding: 0;
            outline: 0;
            border: none;
            width: 100%;
          }

          gux-dropdown input:focus {
            border: none !important;
            box-shadow: none !important;
          }

          gux-dropdown div.gux-dropdown .select-field {
            height: 24px;
          }

          gux-dropdown div.gux-dropdown .select-field .ghost {
            background-color: transparent;
            color: transparent;
            border: none;
          }

          gux-text-field > div > div.gux-field {
            height: 24px;
          }

          gux-dropdown div.gux-dropdown .select-field button.dropdown-indicator {
            display: none;
          }


          .gux-tag-popover-tags-container {
            display: flex;
            flex-wrap: wrap;
          }

          .gux-tag-popover-input-container {
            position: relative;
            flex: 1 1 auto;
          }
        }
        
      }

      &.opened {
        display: block;
      }
    }

    .gux-tag-popover-options {
      display: none;
      position: absolute;
      margin: 0;
      margin-top: 12px;
      padding: 8px 0;
      background: @menu-background-color;
      border: 1px solid @menu-border-color;
      box-shadow: none;
      z-index: 1;
      color: @gux-charcoal-grey;

      &.opened {
        display: block;
      }

      > gux-tag-popover-option {
        position: relative;
        height: 32px;
        line-height: 32px;
        padding: 0 16px;
        display: flex;
        cursor: pointer;

        & {
          &[selected] {
            background: @gux-text-selection;
            color: @gux-charcoal-grey;
          }

          &:hover,
          &:focus {
            background: @gux-dark-blue;
            color: @gux-off-white;
          }
        }
      }
    }
  }
}

// Theming

// Dark
.gux-tag-popover-dark-theme {
  color: @gux-off-white;

  div.gux-tag-popover {
    .gux-tag-popover-title {
      color: @gux-charcoal-grey;
    }
  }
}

.gux-dark-theme {
  gux-tag-popover-beta {
    .gux-tag-popover-dark-theme();
  }
}

gux-tag-popover-beta.gux-dark-theme {
  .gux-tag-popover-dark-theme();
}

// Light
.gux-tag-popover-light-theme {
  color: @gux-charcoal-grey;
}

.gux-light-theme {
  gux-tag-popover-beta {
    .gux-tag-popover-light-theme();
  }
}

gux-tag-popover-beta.gux-light-theme {
  .gux-tag-popover-light-theme();
}

// Default Theme
gux-tag-popover-beta {
  .gux-tag-popover-light-theme();
}
