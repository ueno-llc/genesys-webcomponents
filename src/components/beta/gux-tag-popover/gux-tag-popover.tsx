import { Component, Element, h, Listen, Prop, State } from '@stencil/core';
import { buildI18nForComponent } from '../../../i18n';
import tagPopoverResources from './i18n/en.json';
import { GuxTag } from '../gux-tag/gux-tag';

@Component({
  styleUrl: 'gux-tag-popover.less',
  tag: 'gux-tag-popover-beta'
})
export class GuxTagPopover {
  @Element()
  root: HTMLElement;
  i18n: (resourceKey: string, context?: any) => string;

  /**
   * Tags color
   */
  @Prop()
  color:
    | 'navy'
    | 'blue'
    | 'electric-purple'
    | 'aqua-green'
    | 'fuscha'
    | 'dark-purple'
    | 'bubblegum-pink'
    | 'olive-green'
    | 'lilac'
    | 'yellow-green';

  /**
   * Indicates the position of the tag button (right or left)
   */
  @Prop()
  position: string = 'left';

  @State()
  opened: boolean;

  @State()
  allOptions: HTMLGuxOptionElement[] = [];

  @State()
  availableOptions: HTMLGuxOptionElement[] = [];

  @State()
  tags: string[] = [];

  @State()
  inputValue: string = '';

  @Listen('deleteTag')
  onDeleteTag(event): void {
    const tags = this.tags;
    tags.splice(event.detail, 1);
    this.filterOptions();
    this.tags = [...tags];
  }

  @Listen('change') // This is not a native "change" event
  handleChange(event) {
    event.stopPropagation();
    this.tags = [...this.tags, event.target.value];
    this.filterOptions();
    event.target.value = this.inputValue = ' ';
    setTimeout(() => {
      event.target.value = this.inputValue = '';
      const dropdownInput = this.root.querySelector(
        'gux-dropdown input'
      ) as HTMLElement;
      dropdownInput.focus();
    });
  }

  async componentWillLoad() {
    this.i18n = await buildI18nForComponent(this.root, tagPopoverResources);
    this.allOptions = Array.from(this.root.querySelectorAll('gux-option'));
    this.availableOptions = [...this.allOptions];
  }

  render() {
    return (
      <div class={`gux-tag-popover ${this.position}`}>
        <div class="gux-tag-popover-container">
          <gux-button
            class={`gux-tag-popover-button ${
              this.tags.length ? 'has-tags' : ''
            }`}
            onClick={() => this.handlerClickOnMenuButton()}
          >
            <gux-icon decorative iconName="ic-tag"></gux-icon>
          </gux-button>
        </div>
        <div class={`gux-tag-popover-menu ${this.opened ? 'opened' : ''}`}>
          <div class="arrow" />
          <div class="gux-tag-popover-menu-container">
            <div class="gux-tag-popover-options-container">
              <b class="gux-tag-popover-title">{this.i18n('menuTitle')}</b>
              <div class="gux-tag-popover-chip-container">
                {this.getTags()}
                {this.getDropdown()}
                <span style={{ display: 'none' }}>
                  <slot />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }

  private getDropdown(): HTMLElement {
    return (
      <div class="gux-tag-popover-input-container">
        <gux-dropdown
          value={this.inputValue}
          onKeyDown={e => this.addCustomTag(e)}
          filterable
        >
          {this.availableOptions.map(option => {
            return <gux-option text={option.text}>{option.text}</gux-option>;
          })}
        </gux-dropdown>
      </div>
    );
  }

  private getTags(): GuxTag[] {
    const tags = [];
    this.tags.map((tag, index) => {
      const tagChip = (
        <gux-tag-beta color={this.color} tagId={index.toString()}>
          {tag}
        </gux-tag-beta>
      );
      tags.push(tagChip);
    });
    return tags;
  }

  private addCustomTag(event): void {
    const target = event.target as HTMLInputElement;
    const value = target.value;

    if (event.key === ' ' || event.key === 'Enter') {
      event.preventDefault();
      if (this.tags.indexOf(value) !== -1) {
        return;
      }
      if (value && value.trim().length) {
        this.tags = [...this.tags, value];
      }
      event.target.value = this.inputValue = ' ';
      setTimeout(() => {
        event.target.value = this.inputValue = '';
      });
    }
  }

  private filterOptions(): void {
    this.availableOptions = this.allOptions.filter(
      option => !this.tags.includes(option.text)
    );
  }

  private handlerClickOnMenuButton(): void {
    this.opened = !this.opened;
  }
}
