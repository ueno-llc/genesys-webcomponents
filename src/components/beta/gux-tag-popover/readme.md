# gux-tag-popover

<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description                                              | Type                                                                                                                                                   | Default     |
| ---------- | ---------- | -------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------ | ----------- |
| `color`    | `color`    | Tags color                                               | `"aqua-green" \| "blue" \| "bubblegum-pink" \| "dark-purple" \| "electric-purple" \| "fuscha" \| "lilac" \| "navy" \| "olive-green" \| "yellow-green"` | `undefined` |
| `position` | `position` | Indicates the position of the tag button (right or left) | `string`                                                                                                                                               | `'left'`    |


## Dependencies

### Depends on

- [gux-button](../../stable/gux-button)
- [gux-icon](../../stable/gux-icon)
- [gux-dropdown](../../stable/gux-dropdown)
- [gux-option](../../stable/gux-dropdown/gux-option)
- [gux-tag-beta](../gux-tag)

### Graph
```mermaid
graph TD;
  gux-tag-popover-beta --> gux-button
  gux-tag-popover-beta --> gux-icon
  gux-tag-popover-beta --> gux-dropdown
  gux-tag-popover-beta --> gux-option
  gux-tag-popover-beta --> gux-tag-beta
  gux-dropdown --> gux-text-field
  gux-dropdown --> gux-icon
  gux-text-field --> gux-icon
  gux-option --> gux-text-highlight
  gux-tag-beta --> gux-icon
  style gux-tag-popover-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
