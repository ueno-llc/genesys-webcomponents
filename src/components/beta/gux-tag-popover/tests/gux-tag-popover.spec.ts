import { newSpecPage } from '@stencil/core/testing';
import { GuxTagPopover } from '../gux-tag-popover';

describe('gux-tag-popover', () => {
  it('should render', async () => {
    const page = await newSpecPage({
      components: [GuxTagPopover],
      html: `
        <gux-tag-popover-beta lang="en">
          <gux-option>English</gux-option>
          <gux-option>Dutch</gux-option>
        </gux-tag-popover-beta>
      `,
      language: 'en'
    });

    expect(page.root).toMatchSnapshot();
  });
});
