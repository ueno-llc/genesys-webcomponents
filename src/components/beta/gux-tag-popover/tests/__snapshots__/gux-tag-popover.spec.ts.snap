// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-tag-popover should render 1`] = `
<gux-tag-popover-beta lang="en">
  <!---->
  <div class="gux-tag-popover left">
    <div class="gux-tag-popover-container">
      <gux-button class="gux-tag-popover-button">
        <gux-icon decorative="" iconname="ic-tag"></gux-icon>
      </gux-button>
    </div>
    <div class="gux-tag-popover-menu">
      <div class="arrow"></div>
      <div class="gux-tag-popover-menu-container">
        <div class="gux-tag-popover-options-container">
          <b class="gux-tag-popover-title">
            Add Tags
          </b>
          <div class="gux-tag-popover-chip-container">
            <div class="gux-tag-popover-input-container">
              <gux-dropdown filterable="" value="">
                <gux-option></gux-option>
                <gux-option></gux-option>
              </gux-dropdown>
            </div>
            <span style="display: none;">
              <gux-option>
                English
              </gux-option>
              <gux-option>
                Dutch
              </gux-option>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</gux-tag-popover-beta>
`;
